Snowflake.Data.Tests/IntegrationTests/ConnectionMultiplePoolsIT.cs

```diff
@@ -95,6 +97,76 @@ public static void AfterAllTests()
             Assert.AreEqual(ConnectionState.Closed, conn4.State);
         }
+
+        [Test]
+        public void TestWaitForTheIdleConnectionWhenExceedingMaxConnectionsLimit()
+        {
+            // arrange
+            var connectionString = ConnectionString + "application=TestWaitForTheIdleConnectionWhenExceedingMaxConnectionsLimit";
+            var pool = SnowflakeDbConnectionPool.GetPool(connectionString);
+            Assert.AreEqual(0, pool.GetCurrentPoolSize(), "expecting pool to be empty");
+            pool.SetMaxPoolSize(2);
+            pool.SetWaitingTimeout(1000);
+            var conn1 = OpenedConnection(connectionString);
+            var conn2 = OpenedConnection(connectionString);
+            var watch = new Stopwatch();
+            
+            // act
+            watch.Start();
+            var thrown = Assert.Throws<SnowflakeDbException>(() => OpenedConnection());
+            watch.Stop();
+
+            // assert
+            Assert.That(thrown.Message, Does.Contain("Unable to connect. Could not obtain a connection from the pool within a given timeout"));
+            Assert.GreaterOrEqual(watch.ElapsedMilliseconds, 1000);
+            Assert.LessOrEqual(watch.ElapsedMilliseconds, 1500);
+            Assert.AreEqual(pool.GetCurrentPoolSize(), 2);
+
+            // cleanup
+            conn1.Close();
+            conn2.Close();
+        }
+
+        [Test]
+        public void TestWaitInAQueueForAnIdleSession()
+        {
+            // arrange
+            var connectionString = ConnectionString + "application=TestWaitInAQueueForAnIdleSession";
+            var pool = SnowflakeDbConnectionPool.GetPool(connectionString);
+            Assert.AreEqual(0, pool.GetCurrentPoolSize(), "the pool is expected to be empty");
+            pool.SetMaxPoolSize(2);
+            pool.SetWaitingTimeout(3000);
+            var threads = new ConnectingThreads(connectionString)
+                .NewThread("A", 0, 2000, true)
+                .NewThread("B", 50, 2000, true)
+                .NewThread("C", 100, 0, true)
+                .NewThread("D", 150, 0, true);
+            var watch = new Stopwatch();
+            
+            // act
+            watch.Start();
+            threads.StartAll().JoinAll();
+            watch.Stop();
+
+            // assert
+            var events = threads.Events().ToList();
+            Assert.AreEqual(4, events.Count);
+            CollectionAssert.AreEqual(
+                new[]
+                {
+                    Tuple.Create("A", "CONNECTED"),
+                    Tuple.Create("B", "CONNECTED"),
+                    Tuple.Create("C", "CONNECTED"),
+                    Tuple.Create("D", "CONNECTED")
+                },
+                events.Select(e => Tuple.Create(e.ThreadName, e.EventName)));
+            Assert.LessOrEqual(events[0].Duration, 1000);
+            Assert.LessOrEqual(events[1].Duration, 1000);
+            Assert.GreaterOrEqual(events[2].Duration, 2000);
```

Comment:
I used Is.InRange(...) for time bounderies.
I don't like the idea of expressing all these constants as variables, because some of them are based on estimations, the formula for some of them could be quite a complicated computation. I think that this version is simpler.